hypertrace-trace-enricher/trace-reader/src/main/java/org/hypertrace/trace/reader/entities/TraceEntityReader.java

```diff
@@ -0,0 +1,32 @@
+package org.hypertrace.trace.reader.entities;
+
+import io.reactivex.rxjava3.core.Maybe;
+import io.reactivex.rxjava3.core.Single;
+import java.util.Map;
+import org.hypertrace.core.attribute.service.cachingclient.CachingAttributeClient;
+import org.hypertrace.core.datamodel.Entity;
+import org.hypertrace.core.datamodel.Event;
+import org.hypertrace.core.datamodel.StructuredTrace;
+import org.hypertrace.entity.data.service.rxclient.EntityDataClient;
+import org.hypertrace.entity.type.service.rxclient.EntityTypeClient;
+import org.hypertrace.trace.reader.attributes.TraceAttributeReader;
+
+public interface TraceEntityReader {
+
+  Maybe<Entity> getAssociatedEntityForSpan(String entityType, StructuredTrace trace, Event span);
+
+  Single<Map<String, Entity>> getAssociatedEntitiesForSpan(StructuredTrace trace, Event span);
+
+  static TraceEntityReader build(
```

Comment:
In general I don't feel like interfaces should dictate constructors or builders. Mostly because if we need to add a new attribute that is going to be a breaking change. Maybe it is better to turn this interface into an abstract class (which will be still a contract).

Reply:
I'd love to minimize any breakable surfaces, but I'm not sure I follow how an abstract class helps this - changing that would still be a breaking change, and we ultimately need to expose something accepting these dependencies in the public api - and I would like to avoid doing that through any implementation class. If the feeling is more that it's mixing the role of an interface, that I could certainly buy - what about moving this into separate builder class similar to https://grpc.github.io/grpc-java/javadoc/io/grpc/ManagedChannelBuilder.html ?